evm_arithmetization/src/cpu/kernel/interpreter.rs

```diff
@@ -524,6 +548,145 @@ impl<F: RichField> State<F> for Interpreter<F> {
+                            log::warn!("There is a wrong opcode on the Kernel side but no error message in the struct logs.");
+                        }
+                    }
+                }
+
+                // Check pc.
+                let txn_pc = cur_txn_struct_logs.pc;
+                if txn_pc != self.get_registers().program_counter as u64 {
+                    log::warn!(
+                        "Wrong pc: it is {} in the Kernel but {} in the struct logs",
+                        self.get_registers().program_counter,
+                        txn_pc
+                    );
+                    return Err(ProgramError::StructLogDebuggerError);
+                }
+
+                // Check stack.
+                if let Some(txn_stack) = cur_txn_struct_logs.stack {
+                    let cur_stack = self.get_full_stack();
+                    let txn_stack = txn_stack
+                        .into_iter()
+                        .map(|s| U256(*s.as_limbs()))
+                        .collect::<Vec<_>>();
+
+                    if txn_stack != cur_stack {
+                        log::warn!(
+                            "Wrong stack: it is {:?} in the Kernel but {:?} in
+                    the struct logs.",
+                            cur_stack,
+                            txn_stack
+                        );
+                        return Err(ProgramError::StructLogDebuggerError);
+                    }
+                };
+            };
+        }
+        Ok(())
+    }
+
+    fn check_against_struct_logs_after_op(
+        &mut self,
+        res: &Result<Operation, ProgramError>,
+        consumed_gas: u64,
+        is_user_mode: bool,
+    ) -> Result<(), ProgramError> {
+        if let Some(struct_logs) = &self.struct_logs
+            && is_user_mode
+            && log_enabled!(log::Level::Debug)
+        {
+            let txn_idx = self.generation_state.next_txn_index;
+            // First, update the gas.
+            self.struct_log_debugger_info.gas -= consumed_gas as usize;
+
+            if let Some(txn_struct_logs) = &struct_logs[txn_idx - 1] {
+                let cur_txn_struct_logs =
+                    txn_struct_logs[self.struct_log_debugger_info.counter].clone();
+                let cur_txn_struct_logs_e = cur_txn_struct_logs.error;
+                if cur_txn_struct_logs_e.is_some() && res.is_err() {
+                    log::warn!("The kernel
+                transaction and struct_logs disagree on error. The struct logs return an errorwhile the Kernel doesn't.");
+                    return Err(ProgramError::StructLogDebuggerError);
+                }
+
+                if res.is_err() && cur_txn_struct_logs_e.is_none() {
+                    // This is not necessarily a discrepancy: the struct logs might not hold
+                    // an error message in case of an error.
+                    log::warn!("There is an error on the Kernel side but no error message in the struct logs.");
```

Comment:
I think this may be confusing, eg if we hit an OufOfGas issue we'd get a warning but this is a _valid_ error that will trigger reversion

Reply:
I noticed that on block 678 (which is now passing), there's an error on the Kernel side but not on the structlogs side. I am therefore not sure whether the structlogs really always return something when there's an error...